diff --git a/src/apprt/gtk/class/window.zig b/src/apprt/gtk/class/window.zig
index c01cad618..d476519e3 100644
--- a/src/apprt/gtk/class/window.zig
+++ b/src/apprt/gtk/class/window.zig
@@ -465,6 +465,7 @@ pub const Window = extern struct {
 
         // Add the page and select it
         const page = tab_view.insert(tab.as(gtk.Widget), position);
+        setTabPageLiveThumbnail(page);
         tab_view.setSelectedPage(page);
 
         // Create some property bindings
@@ -503,6 +504,17 @@ pub const Window = extern struct {
         return page;
     }
 
+    fn setTabPageLiveThumbnail(page: *adw.TabPage) void {
+        var value = gobject.ext.Value.newFrom(true);
+        defer value.unset();
+
+        gobject.Object.setProperty(
+            page.as(gobject.Object),
+            "live-thumbnail",
+            &value,
+        );
+    }
+
     pub const SelectTab = union(enum) {
         previous,
         next,
